src/core/tsi/ssl_transport_security.cc

```diff
@@ -1805,8 +1888,11 @@ tsi_result tsi_create_ssl_client_handshaker_factory_with_options(
     return TSI_INVALID_ARGUMENT;
   }
 
-#if defined(OPENSSL_NO_TLS1_2_METHOD) || OPENSSL_API_COMPAT >= 0x10100000L
+#if OPENSSL_VERSION_NUMBER >= 0x10100000
   ssl_context = SSL_CTX_new(TLS_method());
+  result = tsi_set_min_and_max_tls_versions(
+      ssl_context, options->min_tls_version, options->max_tls_version);
+  if (result != TSI_OK) return result;
```

Comment:
TLSv1_2_method() is a deprecated method. Let's remove it here.
We need to change tsi_set_min_and_max_tls_versions() to be no-op if OPENSSL_VERSION_NUMBER < 0x10100000.

It is possible that gRPC user using an old version of OpenSSL/BoringSSL (e.g., 1.0.x) may get downgrade attack, if both client and server use old version of SSL library. But OpenSSL 1.1 has been released for 3+ years and TLSv1_2_method() is a depreciated method. I recommend removing it. 

@davidben what is your recommendation?

Reply:
If it is safe to remove `TLSv1_2_method()`, that's great and I'll do so. In that case, should we just remove the checks for OpenSSL < 1.1 entirely?

I'll leave it up with @jiangtaoli2016 and @davidben as to whether we abandon OpenSSL < 1.1 support.

If we do leave the checks for OpenSSL < 1.1, then I don't think `tsi_set_min_and_max_tls_versions` should be no-op, I think it should return an error in that case. (This is what the current implementation does - it goes right to the default case in the switch statement.) @jiangtaoli2016 WDYT?

Reply:
I suggest that we remove `#if OPENSSL_VERSION_NUMBER >= 0x10100000` in line 1891 and remove deprecated TLSv1_2_method().

By doing so, In `tsi_set_min_and_max_tls_versions()`, if application uses OpenSSL 1.0.x, it will fall into Line 914 and fail. This will makes any OpenSSL 1.0.x user cannot use gRPC. That is not what we want.

Thus I suggest the following:
```
 ssl_context = SSL_CTX_new(TLS_method());
#if OPENSSL_VERSION_NUMBER >= 0x10100000
  result = tsi_set_min_and_max_tls_versions(
      ssl_context, options->min_tls_version, options->max_tls_version);
  if (result != TSI_OK) return result;
#endif
```
You can also remove `#if OPENSSL_VERSION_NUMBER >= 0x10100000` inside tsi_set_min_and_max_tls_versions()

Reply:
I'm in agreement on the changes to `tsi_set_min_and_max_tls_versions()`.

However, if we remove `TLSv1_2_method`, then an OpenSSL 1.0 user might regress to using TLS 1.0 or 1.1, which would be bad. If we want to support OpenSSL 1.0 users, then I think we need to still use `TLSv1_2_method`. (@davidben Please correct me if I'm wrong.)

Reply:
Per an offline discussion with @jiangtaoli2016, I've made the following changes:

- Calling `tsi_set_min_and_max_tls_versions` is a no-op when using OpenSSL versions < 1.1.
- Removed the use of `TLSv1_2_method`. This API is deprecated.

Since the API `SSL_CTX_set_min_proto_version` requires OpenSSL >= 1.1, a user of OpenSSL 1.0 could regress to using TLS 1.0 or 1.1 when the `TLS_method` is called. This is acceptable because OpenSSL 1.0.2 is now end-of-life.

Reply:
Unless you all have set up a compatibility function, I don't think OpenSSL 1.0.2 even has `TLS_method`. It spells it `SSLv23_method`, which, depending on build options, could also enable SSL 2 and SSL 3. If you're going to be adding an ifdef to smooth that over, it's probably about as easy to use `TLSv1_2_method`.

Reply:
Thanks for catching this David - I've added back `TLSv1_2_method` when using OpenSSL < 1.1.